File: .opencode/tools/notion-publish.ts
import { tool } from "@opencode-ai/plugin";
import path from "path";
import fs from "fs";

/**
 * notion_publish — create a Notion page and write VFP markdown content to it.
 *
 * Calls notion-publish.py via stdin for the markdown content (avoids
 * shell-escaping issues with large multi-line strings).
 *
 * Returns the page URL on success.
 *
 * Script resolution order:
 *   1. .opencode/tools/notion-publish.py (project-level, works in CI)
 *   2. ~/.config/vfp-agent/tools/notion-publish.py (global install)
 */
export default tool({
  description:
    "Publish a VFP to Notion. Creates a page under parent_page_id with the " +
    "given title and writes the full markdown content using the Notion Markdown API " +
    "(### headings produce real heading_3 blocks). Returns the page URL.",
  args: {
    parent_page_id: tool.schema
      .string()
      .describe("ID of the parent Notion page (from notion_find_parent)"),
    title: tool.schema
      .string()
      .describe('Page title, e.g. "VFP — add booking system"'),
    markdown: tool.schema
      .string()
      .describe("Full VFP content in markdown format"),
  },
  async execute(args, _context) {
    const projectScript = path.join(process.cwd(), ".opencode", "tools", "notion-publish.py");
    const globalScript = path.join(
      process.env.HOME ?? "~",
      ".config/vfp-agent/tools/notion-publish.py"
    );
    const script = fs.existsSync(projectScript) ? projectScript : globalScript;
    const result =
      await Bun.$`echo ${args.markdown} | python3 ${script} --parent-id ${args.parent_page_id} --title ${args.title}`.text();
    return result.trim();
  },
});

File: .opencode/tools/notion-publish.py
#!/usr/bin/env python3
"""
notion-publish — create a Notion page and write VFP content to it.

Inputs:
  --parent-id   required — ID of the parent Notion page
  --title       required — page title (e.g. "VFP — add booking system")
  stdin         required — full VFP markdown content

Environment:
  NOTION_TOKEN  required — Notion Personal Access Token

Output:
  stdout — page URL on success
  stderr — error message on failure
  exit 0 on success, exit 1 on failure

Markdown format:
  ### headings produce real heading_3 blocks (via Notion Markdown API)
  - bullets produce bulleted_list_item blocks
  Plain paragraphs produce paragraph blocks
"""
import argparse
import json
import os
import sys
import urllib.request


def notion(token, method, path, body=None, version='2022-06-28'):
    req = urllib.request.Request(
        f'https://api.notion.com/v1/{path}',
        data=json.dumps(body).encode() if body is not None else None,
        headers={
            'Authorization': f'Bearer {token}',
            'Content-Type': 'application/json',
            'Notion-Version': version,
        },
        method=method,
    )
    return json.loads(urllib.request.urlopen(req).read())


def main():
    parser = argparse.ArgumentParser(description='Publish a VFP to Notion')
    parser.add_argument('--parent-id', required=True, help='Parent page ID')
    parser.add_argument('--title', required=True, help='Page title')
    args = parser.parse_args()

    token = os.environ.get('NOTION_TOKEN', '').strip()
    if not token:
        print('ERROR: NOTION_TOKEN is not set', file=sys.stderr)
        sys.exit(1)

    markdown = sys.stdin.read()
    if not markdown.strip():
        print('ERROR: no markdown content on stdin', file=sys.stderr)
        sys.exit(1)

    # Step 1 — create the page (properties only, no content)
    try:
        page = notion(token, 'POST', 'pages', {
            'parent': {'page_id': args.parent_id},
            'properties': {
                'title': {'title': [{'text': {'content': args.title}}]},
            },
        })
    except Exception as e:
        print(f'ERROR: failed to create page: {e}', file=sys.stderr)
        sys.exit(1)

    page_id = page['id']
    page_url = page['url']

    # Step 2 — write content via Markdown API (### → real heading_3 blocks)
    try:
        notion(token, 'PATCH', f'pages/{page_id}/markdown',
               {'type': 'replace_content', 'replace_content': {'new_str': markdown}},
               version='2026-03-11')
    except Exception as e:
        print(f'ERROR: failed to write content to page {page_url}: {e}', file=sys.stderr)
        sys.exit(1)

    print(page_url)


if __name__ == '__main__':
    main()
